Change IndexAnalyzers default analyzer access

cbuescher opened the pull request:
Currently IndexAnalyzers keeps the three default analyzers as separate class members
although they should refer to the same analyzers held in the additional
analyzers map under the default names. This assumption should be made more
explicit by keeping all analyzers in the map. This change adapts the constructor
to check all the default entries are there and the getters to reach into the map
with the default names when needed.

Commit:
Centralize IndexAnalyzers default analyzer access

Currently IndexAnalyzers keeps the three default as separate class members
although they should refer to the same analyzers held in the additional
analyzers map under the default names. This assumption should be made more
explicit by keeping all analyzers in the map. This change adapts the constructor
to check all the default entries are there and the getters to reach into the map
with the default names when needed.

elasticmachine commented:
Pinging @elastic/es-search

cbuescher commented:
@romseygeek just chatted about this, mind to take a look and let me know what you think? It shouldn't change any existing behaviour, just make the class a little easier to use and reason about, e.g. currently close() only closes the analyzers in the map and we don't check that the three default references point to entries in the map, so they might accidentally be missed out.

romseygeek commented on test/framework/src/main/java/org/elasticsearch/index/engine/TranslogHandler.java:
```diff
@@ -60,9 +61,13 @@ long appliedOperations() {
+                new NamedAnalyzer("default", AnalyzerScope.INDEX, new StandardAnalyzer()));
+        analyzers.put(AnalysisRegistry.DEFAULT_SEARCH_QUOTED_ANALYZER_NAME,
+                new NamedAnalyzer("default", AnalyzerScope.INDEX, new StandardAnalyzer()));
+        IndexAnalyzers indexAnalyzers = new IndexAnalyzers(indexSettings, analyzers, emptyMap(), emptyMap());
```
Rather than always requiring all three defaults to be contained in the Map, maybe it would be nicer to have some fallback logic in `getDefaultSearchAnalyzer` and `getDefaultSearchQuoteAnalyzer` that returns the default index analyzer if nothing specific has been added?

It would also be helpful to have an IndexAnalyzers constructor that takes a single analyzer and returns that as the default.

romseygeek replied:
niit: You only need `default` here now

romseygeek reviewed (changes requested):
Thanks @cbuescher - I left one comment, I think we can make this still easier to use.

Commit:
Merge branch 'master' into simplify-IndexAnalyzers

Commit:
Improving default handling

cbuescher commented:
@romseygeek thanks, good suggestion. I changed the default handling and added a test. I don't feel like we need an additional convenience constructor  just for the test code that you mentioned though, but maybe I missed some reasons for it so please let me know if you consider it important.

romseygeek commented on server/src/test/java/org/elasticsearch/index/analysis/IndexAnalyzersTests.java:
```diff
@@ -0,0 +1,91 @@
+
+    /**
+     * test the checks in the constructor
+     * @throws IOException
```
nit: I don't think we need this javadoc?

cbuescher replied:
Yeah, I tend to miss short javadocs in tests quite often (especially if it comes to thinking about removing them, then its good to see what a certain test was set up to do) but in this case its quite obvious.

romseygeek reviewed (approved):
Two minor nitpicks, but LGTM otherwise - no need for another review.

Commit:
iter
